### PR TITLE
priviege, executor: add set role option

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve? 
Support SQL like:
```
SET ROLE ALL
SET ROLE DEFAULT
SET ROLE ALL EXCEPT role [, role ]
SET ROLE NONE
```

### What is changed and how it works?
Reading roles information from privilege cache and change `ActiveRole` session variable.
adding `GetAllRole` interface for Manger, which could get roles from cache.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
**IGNORE** this comment~

/run-common-test tidb-test=pr/579
/run-integration-common-tests tidb-test=pr/579

**Comment by imtbkcat** [comment]
PTAL @tiancaiamao

File: executor/simple.go [hunk 1]
```diff
@@ -254,6 +255,89 @@ func (e *SimpleExec) executeSetRole(s *ast.SetRoleStmt) error {
...
+	// Deal with SQL like `SET ROLE ALL EXCEPT role1, role2;`
+	for _, r := range s.RoleList {
+		if r.Hostname == "" {
+			r.Hostname = "%"
+		}
+	}
+	checker := privilege.GetPrivilegeManager(e.ctx)
+	user, host := e.ctx.GetSessionVars().User.AuthUsername, e.ctx.GetSessionVars().User.AuthHostname
+	roles, afterExcept := checker.GetAllRoles(user, host), make([]*auth.RoleIdentity, 0)
+	for _, r := range roles {
```
Comment (tiancaiamao) [inline executor/simple.go]: It remind me of functional programming...

```
func filter(arr []*auth.RoleIdentity, fn f(*auth.RoleIdentity) bool) []*auth.RoleIdentity {
     i, j := 0,0
     for i:=0; i<len(arr); i++ {
         if f(arr[i]) {
            arr[j] = arr[i]
            j++
         }
     }
     return arr[:j]
}

banned := func(*auth.RoleIdentity) bool {
          for _, ban := range s.RoleList {
              if ban.Hostname == r.Hostname && ban.Username == r.Username {
                 return true
			        }        
          }
          return false
}

filter(roles, not(banned))
```

File: executor/simple.go [hunk 2]
```diff
@@ -254,6 +255,89 @@ func (e *SimpleExec) executeSetRole(s *ast.SetRoleStmt) error {
...
+func (e *SimpleExec) setRoleAllExcept(s *ast.SetRoleStmt) error {
+	// Deal with SQL like `SET ROLE ALL EXCEPT role1, role2;`
+	for _, r := range s.RoleList {
+		if r.Hostname == "" {
+			r.Hostname = "%"
+		}
+	}
+	checker := privilege.GetPrivilegeManager(e.ctx)
+	user, host := e.ctx.GetSessionVars().User.AuthUsername, e.ctx.GetSessionVars().User.AuthHostname
+	roles, afterExcept := checker.GetAllRoles(user, host), make([]*auth.RoleIdentity, 0)
```
Comment (tiancaiamao) [inline executor/simple.go]: make([]*auth.RoleIdentity, 0, len(s.RoleList))

File: executor/simple.go [hunk 3]
```diff
@@ -254,6 +255,89 @@ func (e *SimpleExec) executeSetRole(s *ast.SetRoleStmt) error {
...
+		return ErrRoleNotGranted.GenWithStackByArgs(roleName, u.String())
+	}
+	return nil
+}
+
+func (e *SimpleExec) setRoleNone(s *ast.SetRoleStmt) error {
+	// Deal with SQL like `SET ROLE NONE;`
+	checker := privilege.GetPrivilegeManager(e.ctx)
+	roles := make([]*auth.RoleIdentity, 0)
+	ok, roleName := checker.ActiveRoles(e.ctx, roles)
```
Comment (tiancaiamao) [inline executor/simple.go]: checker.ActiveRoles(e.ctx, nil) ?

File: privilege/privileges/cache.go [hunk 4]
```diff
@@ -940,6 +940,18 @@ func (p *MySQLPrivilege) getDefaultRoles(user, host string) []*auth.RoleIdentity
 	return ret
 }
 
+func (p *MySQLPrivilege) getAllRoles(user, host string) []*auth.RoleIdentity {
+	ret := make([]*auth.RoleIdentity, 0)
+	key := user + "@" + host
+	edgeTable, ok := p.RoleGraph[key]
```
Comment (tiancaiamao) [inline privilege/privileges/cache.go]:
```
if !ok {
    return nil
}

ret := make([]*auth.RoleIdentity, 0, len(edgeTable.roleList))
```

When you're writing Go, remind:
1. test the branch that could return early
2. write cap for make() whenever possible

Comment (imtbkcat) [reply privilege/privileges/cache.go]: fixed @tiancaiamao

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**